### PR TITLE
[PoC] Use RayCollector + weight update extension points with super v0 "param server"

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
The main goal of this PoC is to 
(1) demonstrate integration of RayCollector
(2) Provide example of using weight sync extension points in RayCollector and ensure it works end to end

I have basically not thought about perf and efficient weight sync strategies yet :(

This PR depends on https://github.com/pytorch/rl/pull/2865/files which contains the bulk of the weight sync logic. (That PR is not rebased onto torchrl main yet),
(1) torchrl main deleted `from_vllm` and now uses `vLLMWrapper`
(2) when rebasing and moving to vLLMWrapper I see issues with lazy_stack that I haven't looked into 😢 

Rewards/Successes
Screenshot 2025-03-28 at 10 03 28 AM

Policy age:
Screenshot 2025-03-28 at 10 16 43 AM

## Weight sync
How does the weight sync change in this PR?

On `main` 
- After `steps_before_sync` steps all `vLLMRolloutActor`s will spin and wait for the trainer to update their weights
- After `steps_before_sync` steps, trainer will all gather to rank 0 and broadcast to *ALL* vLLM workers then notify them to continue rollouts

In this PR:
- there is an extra rank in the global process group for the trainer that subclasses `RemoteWeightUpdaterBase` and I call a v0 "param server"
- every `steps_before_sync` steps, trainer workers all gather and rank0 broadcasts to that last rank
- RayCollector utilizes [`_async_iterator`](https://github.com/pytorch/rl/blob/e1d3fd488d48d68999782df4dc58925e40943642/torchrl/collectors/distributed/ray.py#L735) which get a batch from one of the remote collectors (vLLM policy + LLMEnv that contains the dataloader) and afterwards because `update_after_each_batch=True` will try to update weights of that specific remote collector. The "param server" tracks versioning and will only broadcast weights to the vLLM if it has a newer version

An implication is that the broadcast happens 1 param server to 1 worker at a time in this PR. But that is largely due to how `async_iterator` works and will be changed

## Things that are broken
(1) tp_size > 1 (need to modify design to make Remote and Local Weight Updaters aware of each other and able to communicate, or alternatively pass in tp_size to RemoteWeightUpdater at __init__)
(2) logging in vLLMRollout Actors
(3) Tokenizer is changed to hf tokenizer 
(4) I think batch_size > 1  (not grpo_samples) might be broken

## Next Steps

- Make remote collectors responsible for updating themselves rather than update being orchestrated by RayCollector
- Actually make weight sync efficient, consider and prototype different strategies for syncing the weights, resharding etc. Not just rely on vLLM `update_weights` to do sharding

### CODE REVIEW

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 1]
```diff
@@ -491,270 +480,73 @@ def _setup_data(
...
+        num_replicas=num_replicas,
+    )
 
-    def init_weight_update_group(self, master_address, master_port, rank, world_size):
-        from vllm.distributed.parallel_state import get_world_group
+    
+    from torchrl.envs import LLMEnv
+    llmenv = LLMEnv.from_dataloader(
+        dataloader=_dataloader,
+        tokenizer=_tokenizer,
```
Comment (mikaylagawarecki) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:506]: I have to use a hf tokenizer here as LLMEnv will call `batch_encode_plus` which the tune tokenizer doesn't have

File: torchtune/dev/grpo/data.py [hunk 2]
```diff
@@ -71,11 +71,13 @@ def _prepare_sample(self, sample: Mapping[str, Any]) -> Dict[str, Any]:
 
         question = BASE_PROMPT % transformed_sample["question"]
 
-        q_tokens = self._tokenizer.encode(question, add_eos=False)
-        mask = [1 for _ in q_tokens]
+        # q_tokens = self._tokenizer.encode(question, add_eos=False)
+        # mask = [1 for _ in q_tokens]
```
Comment (mikaylagawarecki) [inline torchtune/dev/grpo/data.py:75]: Delegated tokenization to LLMEnv

File: recipes/configs/dev/qwen3B_full_grpo.yaml [hunk 3]
```diff
@@ -114,10 +114,11 @@ num_ref_workers: 1
 num_fsdp_workers: 2
 
 vllm:
-  num_workers: 5
+  num_workers: 4
   tp_size: 1
   batch_size: 1 # TODO: for now, needs to match recipe batch size
-  queue_maxsize: 3 #num workers * steps_before_sync
+  queue_maxsize: 3
+  gpu_memory_utilization: 0.7
```
[the remark this answers is not in the document]
Comment (SalmanMohammadi) [reply recipes/configs/dev/qwen3B_full_grpo.yaml:121]: I also pretty much always OOM when using gpu_memory_utilization > 0.85

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 4]
```diff
@@ -491,270 +480,73 @@ def _setup_data(
...
+    grpo_samples = cfg.grpo_samples
+    batch_size = cfg.batch_size
+    # _tokenizer = config.instantiate(cfg.tokenizer)
+    # collate_name = cfg.get(
+    #     "collate_fn", "torchtune.dev.grpo.data.padded_collate_rl"
+    # )
+    _tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-3B")
+    _tokenizer.pad_token = _tokenizer.eos_token
+    _tokenizer.padding_side = "right"
+    _dataloader = _setup_data(
```
[the remark this answers is not in the document]
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:493]: This intersects with @felipemello1's work. We can lean on him to do this part: we do want VLLM to tokenize, not even the LLMEnv eventually

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 5]
```diff
@@ -368,76 +370,63 @@ def run(self):
...
-    def start_weight_update(self, param_list, policy_version):
-        # Update the policy version when weights are synchronized
-        self.policy_version = policy_version
-        for name, dtype, shape in param_list:
-            self.llm.collective_rpc("update_weight", args=(name, dtype, shape))
-
-    def llm_collective_rpc(self, *args, **kwargs):
-        self.llm.collective_rpc(*args, **kwargs)
+# TODO: logging needs to be put into the Collector somehow
+# class vLLMRolloutActor:
```
[the remark this answers is not in the document]
Comment (Darktex) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:374]: Logging is fine, but I have a strong preference for this to be an actor with its own class rather than a random function as that helps clarity a lot IMHO (we think in actors)

**Review by Darktex — CHANGES_REQUESTED** [review]
I propose this:

- Keep the Tune tokenizer for now in Training
- Make the collector an actual actor
- I'm ok to go back on logging for now, but file an issue and discuss a plan with Felipe
- Work with Felipe on how to not tokenize from a Tune dataset and send the raw strings in
- Also, don't tokenize in the LLMEnv either, just send VLLM the actual strings

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 6]
```diff
@@ -1746,11 +1596,56 @@ def _init_weight_sync_pg(self):
         ray.get(self.actor_workers[0].set_vllm_engines.remote(self.rollout_workers))
 
     def train(self):
-        rollout_handles = [worker.rollout.remote() for worker in self.rollout_workers]
-        self.rollout_workers[0].print_me.remote("hello vllm worker, it's __main__")
+        from transformers import AutoTokenizer
+
+        tokenizer = AutoTokenizer.from_pretrained("Qwen/Qwen2.5-3B")
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:1601]: This tokenizer is not used, I just use it to get the pad_token_id

File: scripts/runnable_recipe_ray_vllm_weight_sync.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (felipemello1) [reply scripts/runnable_recipe_ray_vllm_weight_sync.py:493]: The dataset should be flexible enough for that. Basically a dataset will take a set of transforms:

message_transform: go from raw_text -> message
tokenizer: message -> encoded
recipe_transform: Encoded -> dictionary/masking

They can be empty, so we can define something like:
```
dataset(message_transform=None, tokenizer=None, recipe_transform=vllm_actor)
```

**Comment by mikaylagawarecki** [comment]
Closing in favor of #29 as we decided that RayCollector orchestrating multiple collectors was too opaque
